barretenberg/cpp/src/barretenberg/vm/avm_trace/AvmMini_deserialization.cpp

```diff
@@ -0,0 +1,176 @@
+ *        checks that each opcode value is in the defined range and extracts the operands
+ *        for each opcode based on the specification from OPCODE_WIRE_FORMAT.
+ *
+ * @param bytecode The bytecode to be parsed as a vector of bytes/uint8_t
+ * @throws runtime_error exception when the bytecode is invalid.
+ * @return Vector of instructions
+ */
+std::vector<Instruction> Deserialization::parse(std::vector<uint8_t> const& bytecode)
+{
+    std::vector<Instruction> instructions;
+    size_t pos = 0;
+    const auto length = bytecode.size();
+
+    while (pos < length) {
+        const uint8_t opcode_byte = bytecode.at(pos);
+
+        if (!Bytecode::is_valid(opcode_byte)) {
+            throw_or_abort("Invalid opcode byte: " + std::to_string(opcode_byte) +
+                           " at position: " + std::to_string(pos));
+        }
+        pos++;
+
+        auto const opcode = static_cast<OpCode>(opcode_byte);
+        std::vector<OperandType> inst_format;
+
+        if (opcode == OpCode::SET) {
+            if (pos == length) {
+                throw_or_abort("Operand for SET opcode is missing at position " + std::to_string(pos));
+            }
+            uint8_t set_tag_u8 = bytecode.at(pos);
+            if (!std::set<uint8_t>{ static_cast<uint8_t>(AvmMemoryTag::U8),
+                                    static_cast<uint8_t>(AvmMemoryTag::U16),
+                                    static_cast<uint8_t>(AvmMemoryTag::U32),
+                                    static_cast<uint8_t>(AvmMemoryTag::U64),
+                                    static_cast<uint8_t>(AvmMemoryTag::U128) }
+                     .contains(set_tag_u8)) {
+                throw_or_abort("Instruction tag for SET opcode is invalid at position " + std::to_string(pos) +
+                               " value: " + std::to_string(set_tag_u8));
+            }
+
+            auto in_tag = static_cast<AvmMemoryTag>(set_tag_u8);
+            switch (in_tag) {
+            case AvmMemoryTag::U8:
+                inst_format = { OperandType::TAG, OperandType::UINT8, OperandType::UINT32 };
+                break;
+            case AvmMemoryTag::U16:
+                inst_format = { OperandType::TAG, OperandType::UINT16, OperandType::UINT32 };
+                break;
+            case AvmMemoryTag::U32:
+                inst_format = { OperandType::TAG, OperandType::UINT32, OperandType::UINT32 };
+                break;
+            case AvmMemoryTag::U64:
+                inst_format = { OperandType::TAG, OperandType::UINT64, OperandType::UINT32 };
+                break;
+            case AvmMemoryTag::U128:
+                inst_format = { OperandType::TAG, OperandType::UINT128, OperandType::UINT32 };
+                break;
+            default: // This branch is guarded above.
+                inst_format = {};
+            }
+        } else {
+            inst_format = OPCODE_WIRE_FORMAT.at(opcode);
```

Comment:
Is the .at() guaranteed to succeed?

Reply:
Ultimately yes, the guard is the opcode validation above. It might fail now, because the OPCODE_WIRE_FORMAT is not defined yet for all opcodes.